testing/simapp/app.go

```diff
@@ -315,22 +321,35 @@ func NewSimApp(
 		&stakingKeeper, govRouter,
 	)
 
-	// Create Transfer Keepers
+	app.IBCFeeKeeper = ibcfeekeeper.NewKeeper(appCodec, keys[ibcfeetypes.StoreKey], app.GetSubspace(ibcfeetypes.ModuleName),
+		app.IBCKeeper.ChannelKeeper, &app.IBCKeeper.PortKeeper,
+	)
+
+	// Create Transfer Keeper and pass IBCFeeKeeper as expected Channel and PortKeeper
+	// since fee middleware will wrap the IBCKeeper for underlying application.
 	app.TransferKeeper = ibctransferkeeper.NewKeeper(
 		appCodec, keys[ibctransfertypes.StoreKey], app.GetSubspace(ibctransfertypes.ModuleName),
-		app.IBCKeeper.ChannelKeeper, &app.IBCKeeper.PortKeeper,
+		app.IBCFeeKeeper, &app.IBCFeeKeeper,
```

Comment:
Question: why do we need to pass the `IBCFeeKeeper` here? In the case that we want to use multiple middleware, what would this look like?

Reply:
Good question.

We need to ensure that all communication from transfer to IBC goes through the fee keeper first. This ensures that the fee keeper can do its logic first before passing on the information to IBC

So that's why IBCFeeKeeper wraps the channel/port keeper methods in `keeper.go`.

In the case of multiple middleware, we again want to ensure the line of communication from the base application through the lowest middleware up the chain to the highest middleware and then to IBC.

Suppose we had the following stack: `fee -> whitelist -> transfer`.

transfer would be given the `whitelist` keeper as its "wrapped" channel and port keeper.
Meanwhile `whitelist` would be given the `fee` keeper as its "wrapped" channel and port keeper
Fee will get the true channel and port keeper.

This ensures that any flow from transfer to IBC. e.g. a `SendPacket` will first go through `whitelist`, then it will go through `fee`, and only then will it reach IBC. At each step, the middleware can perform its custom logic at the right step of the chain, and there is no way for a middleware/app at one part of the stack to "bypass" the higher levels of middleware and communicate with IBC directly

Reply:
This should be written up in the middleware developer guide

Reply:
Nice explanation @AdityaSripal 

As a side note, the type being passed in is `types.ChannelKeeper` (from expected_interfaces.go) not `feekeeper.Keeper`

Reply:
Thanks for the explanation. :+1: